.release-notes/4173.md

```diff
@@ -0,0 +1,18 @@
+## Fix for segfault when using Array.copy_to() with an empty Array
+
+Currently, the `Array.copy_to()` function does not check whether the Array is empty or has been instanced with any values. This leads to a runtime crash when calling `copy_to()` on an empty array such as in this example (issue #4172):
+
+```pony
+class Foo
+    embed _array: Array[I32] = []
+
+new create() => None
+new clone(src: Foo) => src._array.copy_to(_array, 0, 0, 10)
+
+actor Main
+
+    new create(env': Env) =>
+        let f2 = Foo.clone(Foo)
+```
+
```

Comment:
```suggestion

`Array` is part of the `builtin` package of the standard library. Only code in `builtin` is allowed to do direct pointer manipulations like those that `Array` uses to implement `copy_to`. Direct pointer manipulations, if done incorrectly, can lead to segfaults and other memory related errors. Because pointer manipulation can be dangerous, only code in the `builtin` package is allowed to do it. 

We consider unsafe pointer operations in the `builtin` package to be of utmost importance and once this bug was found moved quickly to release a fix.
```